### PR TITLE
(PC-15704) feat(OfflineMode): added offline mode with bookings support

### PR DESCRIPTION
**Pull request by kopax-polyconseil**
Link to JIRA ticket: https://passculture.atlassian.net/browse/PC-15704

## Checklist

I have:

- [x] Made sure the title of my PR follows the [convention](1) `($jira) $type($scope): $summary`.
- [x] Made sure my feature is working on the relevant real / virtual devices (native and web).
- [x] Written **unit tests** native (and web when implementation is different) for my feature.
- [ ] Added a **screenshot** for UI tickets.
- [ ] Attached a **ticket number** or **GitHub dev name** for any added TODO / FIXME.
- [ ] Added new reusable components to **AppComponents** page and communicate to the team on slack.
- [ ] If my PR is a bugfix, I add the link of the "résolution de problème sur le bug" [on Notion](2)
- [ ] Made sure translations file is up-to-date (`yarn translations:extract`)

## Screenshots

| iOS | Android | Mobile - Chrome | Desktop - Chrome | Desktop - Safari |
| --: | ------: | --------------: | ---------------: | ---------------: |
|  image   | image        |         image        |     image             |                  |

[1]: https://github.com/pass-culture/pass-culture-app-native/blob/master/doc/standards/pr-title.md
[2]: https://www.notion.so/passcultureapp/R-solution-de-probl-mes-sur-les-bugs-5dd6df8f6a754e6887066cf613467d0a

### CODE REVIEW

File: src/features/home/pages/Home.tsx [hunk 1]
```diff
@@ -192,6 +194,14 @@ export const Home: FunctionComponent = () => {
   )
 }
 
+export const Home: FunctionComponent = () => {
+  const netInfo = useNetInfo()
+  if (netInfo.isConnected) {
+    return <OnlineHome />
```
[the remark this answers is not in the document]
Comment (kopax-polyconseil) [reply src/features/home/pages/Home.tsx:200]: pareil que https://github.com/pass-culture/pass-culture-app-native/pull/3121#discussion_r916858427

File: src/libs/network/__tests__/OfflineModeContainer.test.tsx [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+jest.mock('libs/network/useNetInfo', () => jest.requireMock('@react-native-community/netinfo'))
+
+const mockUseNetInfo = useNetInfoDefault as jest.Mock
+
+describe('<OfflineModeContainer />', () => {
+  mockUseNetInfo.mockImplementation(() => ({ isConnected: false }))
+
+  it('should render children and show banner when offline', () => {
+    const renderAPI = renderOfflineModeContainer()
+    expect(renderAPI.queryByText('Hello World')).toBeTruthy()
```
[the remark this answers is not in the document]
Comment (kopax-polyconseil) [reply src/libs/network/__tests__/OfflineModeContainer.test.tsx:17]: C'est pour tester que le children render bien, c'est un children render pour le test.

File: src/libs/react-query/tests/usePersistQuery.test.tsx [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+jest.mock('libs/network/useNetInfo', () => jest.requireMock('@react-native-community/netinfo'))
+const mockUseNetInfo = useNetInfoDefault as jest.Mock
+
+type TestData = {
+  id: number
+  description: string
+}
+
+// eslint-disable-next-line local-rules/no-allow-console
+allowConsole({
```
[the remark this answers is not in the document]
Comment (kopax-polyconseil) [reply src/libs/react-query/tests/usePersistQuery.test.tsx:23]: c'est pour 2 tests : `should show offline data first, then online data` et `should return defined when select find the entity (10 ms)`

C'est un contournement, je n'ai pas trouvé comment faire sans.

File: src/features/home/api/useExcluOffer.ts [hunk 4]
```diff
@@ -14,6 +17,6 @@ export const useExcluOffer = (id: number) => {
         return undefined
       }
     },
-    { enabled: typeof id === 'number' }
+    { enabled: !netInfo.isConnected && typeof id === 'number' }
```
[the remark this answers is not in the document]
Comment (kopax-polyconseil) [reply src/features/home/api/useExcluOffer.ts:20]: Thanks !

File: src/features/bookings/components/NoBookingsView.test.tsx [hunk 5]
```diff
@@ -16,8 +17,25 @@ jest.mock('features/search/pages/SearchWrapper', () => ({
   }),
 }))
 
+jest.mock('libs/network/useNetInfo', () => jest.requireMock('@react-native-community/netinfo'))
+
+const mockUseNetInfo = useNetInfoDefault as jest.Mock
+
 describe('<NoBookingsView />', () => {
+  it('should render online no bookings view', () => {
+    mockUseNetInfo.mockReturnValueOnce({ isConnected: true })
```
[the remark this answers is not in the document]
Comment (kopax-polyconseil) [reply src/features/bookings/components/NoBookingsView.test.tsx:26]: C'est écris avec `online` et `offline` mais j'améliore le texte

**Review by Lucasbeneston — APPROVED** [review]

**Pull request merged**